### PR TITLE
[flang][omp] Emit omp.workshare in frontend

### PR DESCRIPTION
**Pull request by ivanradanov**
2/4

1/4 https://github.com/llvm/llvm-project/pull/101443
2/4 https://github.com/llvm/llvm-project/pull/101444
3/4 https://github.com/llvm/llvm-project/pull/101445
4/4 https://github.com/llvm/llvm-project/pull/101446

### CODE REVIEW

**Comment by ivanradanov** [comment]
Should we have a `-use-experimental-workshare` or similar flag to facilitate some temporary in-tree development as this may require more moving pieces.

**Comment by tblah** [comment]
> Should we have a `-use-experimental-workshare` or similar flag to facilitate some temporary in-tree development as this may require more moving pieces?

A flag like that sounds appropriate yes. The current code changes look good.

**Review by tblah — APPROVED** [review]
LGTM, thanks

File: flang/lib/Lower/OpenMP/OpenMP.cpp [hunk 1]
```diff
@@ -1272,6 +1272,15 @@ static void genTaskwaitClauses(lower::AbstractConverter &converter,
       loc, llvm::omp::Directive::OMPD_taskwait);
 }
 
+static void genWorkshareClauses(lower::AbstractConverter &converter,
```
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:1275]: Ultra-nit: Move below `genTeamsClauses` to keep alphabetical sorting.

**Review by skatrak — APPROVED** [review]
Thank you, LGTM as well.

**Pull request merged**